dbms/src/Storages/DeltaMerge/VersionChain/VersionChain.cpp

```diff
@@ -0,0 +1,383 @@
+
+    auto cf_reader = cf.getReader(dm_context, data_provider, getHandleColumnDefinesPtr<HandleType>(), ReadTag::MVCC);
+    auto block = cf_reader->readNextBlock();
+    RUNTIME_CHECK_MSG(
+        cf.getRows() == block.rows(),
+        "ColumnFile<{}> returns {} rows. Read all rows in one block is required!",
+        cf.toString(),
+        block.rows());
+
+    const auto & column = *(block.begin()->column);
+    RUNTIME_CHECK(column.size() > offset, column.size(), offset);
+
+    const auto handle_col = ColumnView<HandleType>(column);
+    auto itr = handle_col.begin() + offset;
+
+    if (calculate_read_packs)
+        calculateReadPacks(itr, handle_col.end());
+
+    replayHandles(dm_context, itr, handle_col.end(), stable_rows, delta_reader);
+    return column.size() - offset;
+}
+
+template <ExtraHandleType HandleType>
+UInt32 VersionChain<HandleType>::replayColumnFileBig(
+    const DMContext & dm_context,
+    const ColumnFileBig & cf_big,
+    const UInt32 stable_rows,
+    const StableValueSpace::Snapshot & stable,
+    const std::span<const ColumnFilePtr> preceding_cfs,
+    DeltaValueReader & delta_reader)
+{
+    auto cf_big_min_max = loadDMFileHandleRange<HandleType>(dm_context, *(cf_big.getFile()));
+    if (!cf_big_min_max) // DMFile is empty.
+        return 0;
+
+    HandleRefType cf_big_min = cf_big_min_max->first;
+    HandleRefType cf_big_max = cf_big_min_max->second;
+
+    // If a ColumnFileBig is ingested by apply snapshot, its preceding should be a ColumnFileDeleteRange.
+    auto is_apply_snapshot = [&]() {
+        if (dmfile_or_delete_range_list.empty())
+            return false;
+        if (auto * delete_range = std::get_if<RowKeyRange>(&dmfile_or_delete_range_list.back()); delete_range)
+            return inRowKeyRange(*delete_range, cf_big_min) && inRowKeyRange(*delete_range, cf_big_max);
+        return false;
+    };
+
+    auto is_dmfile_intersect = [&](const DMFile & file) {
+        auto file_min_max = loadDMFileHandleRange<HandleType>(dm_context, file);
+        if (!file_min_max)
+            return false;
+        const auto & [file_min, file_max] = *file_min_max;
+        return cf_big_min <= file_max && file_min <= cf_big_max;
+    };
+    // If a ColumnFileBig is ingested by ingest sst, it should not intersect with any preceding ColumnFile.
+    // Also, it should not have other ColumnFileTiny, ColumnFileInMemory or ColumnFileDeleteRange.
+    auto is_ingest_sst = [&]() {
+        for (const auto & file : stable.getDMFiles())
+            if (is_dmfile_intersect(*file))
+                return false;
+
+        for (const auto & cf : preceding_cfs)
+        {
+            if (const auto * t = cf->tryToBigFile(); t)
+            {
+                if (is_dmfile_intersect(*(t->getFile())))
+                    return false;
+            }
+            else
+                return false;
+        }
+        return true;
+    };
+
+    if (likely(is_apply_snapshot() || is_ingest_sst()))
```

Comment:
If the ColumnBig doesn't come from the applying snapshot or the ingest sst, then it comes from delta-merge which is not a common case?

Reply:
Do you mean the ColumnFileBig which is created by delta-merge is not called an **`ingest`** here?